### PR TITLE
rejection mechanism for dsp/incentive service

### PR DESCRIPTION
**Pull request by pschuermann97**
## Changes in this PR:

rejection mechanism for transactions:
incentive service now queries dsprotection service whether dsid of occured transaction is already contained in database.
If dsp service is alive and returns true, the transaction is immediately aborted and not ever recorded in the database.

## Checklist:

Select whatever applies:

- [ ] Test written
- [x] Code Properly Formatted
- [x] Documented

### CODE REVIEW

**Comment by this-kramer** [comment]
I think you need adapt the test cases to the new workflow 🤔

File: services/incentive/src/main/java/org/cryptimeleon/incentive/services/incentive/IncentiveService.java [hunk 1]
```diff
@@ -163,7 +164,36 @@ private SideEffect handleSpendRequest(BigInteger promotionId, UUID basketId, UUI
...
+        *    => needs to be invalidated now
+        *
+        * For the invalidation in step 5 to be possible,
+        * it is critical that transactions that occured during dsp service downtime are always recorded in the database,
+        * no matter whether dsid was already known.
+        */
+        GroupElement usedTokenDsid = spendRequest.getDsid();
+        if(offlineDspRepository.dspServiceIsAlive() && offlineDspRepository.containsDsid(usedTokenDsid)) {
+            return new CaughtDoubleSpendingSideEffect("Double-spending attempt detected: " + usedTokenDsid + " has already been spent!");
+        }
```
Comment (JanBobolz) [inline services/incentive/src/main/java/org/cryptimeleon/incentive/services/incentive/IncentiveService.java:193]: The `offlineDspRepository.dspServiceIsAlive()` only checks that the server is not physically down, right? (and it seems like it would throw an unfriendly exception instead of returning `false`).
This should probably be replaced by a check whether or not the store has pretend-lost connection to the database 🤔.

Comment (this-kramer) [reply services/incentive/src/main/java/org/cryptimeleon/incentive/services/incentive/IncentiveService.java:193]: Maybe just let the containsDsid call throw an offline exception triggered by the simulated offline feature

**Comment by pschuermann97** [comment]
Tests work again, added possibility to turn off double-spending database synchronization for test cases that do not intend to test it.
DBSyncIntegrationTest is the test class that tests all functionality around the Spend-Deduct mechanism.

File: services/incentive/src/main/java/org/cryptimeleon/incentive/services/incentive/IncentiveService.java (hunk 1 above)
Comment (JanBobolz) [reply services/incentive/src/main/java/org/cryptimeleon/incentive/services/incentive/IncentiveService.java]: What about this? Did you build in Paul's recommendation?

Comment (pschuermann97) [reply services/incentive/src/main/java/org/cryptimeleon/incentive/services/incentive/IncentiveService.java]: Why would it throw an unfriendly exception? If the blocking call in DsProtectionClient.dspIsAlive() in client package times out, the returned String is null as far as I know. So the method will return false.

File: services/incentive/src/main/java/org/cryptimeleon/incentive/services/incentive/IncentiveService.java [hunk 2]
```diff
@@ -163,7 +176,38 @@ private SideEffect handleSpendRequest(BigInteger promotionId, UUID basketId, UUI
...
+        */
+        if(doSync) {
+            GroupElement usedTokenDsid = spendRequest.getDsid();
+            if(offlineDspRepository.dspServiceIsAlive() && offlineDspRepository.containsDsid(usedTokenDsid)) {
+                return new CaughtDoubleSpendingSideEffect("Double-spending attempt detected: " + usedTokenDsid + " has already been spent!");
+            }
+            else {
+                offlineDspRepository.addToDbSyncQueue(promotionId, tid, spendRequest, spendProviderOutput);
+            }
+        }
```
Comment (this-kramer) [inline services/incentive/src/main/java/org/cryptimeleon/incentive/services/incentive/IncentiveService.java:210]: Not a fan of this: We make the system insecure just to make testing easier. Instead provide a separate implementation of of the DSP-repository interface that is injected in the offlineDspRepository variable through a Bean that does not try to speak with the DSP service

File: services/incentive/src/main/java/org/cryptimeleon/incentive/services/incentive/IncentiveService.java (hunk 1 above)
Comment (JanBobolz) [reply services/incentive/src/main/java/org/cryptimeleon/incentive/services/incentive/IncentiveService.java]: [the documentation](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Mono.html#block-java.time.Duration-) reads to me like it would raise a `RuntimeException`. Maybe I'm missing something there.
([Context](https://github.com/cryptimeleon/incentive-system/blob/424aadfe80c914e74baaa9d8ceaedf69fed0ffaa/services/client/src/main/java/org/cryptimeleon/incentive/client/DSProtectionClient.java#L123)).

But in any case, you should make it so that if we're "simulated offline", then you shouldn't check the dsid.

**Comment by this-kramer** [comment]
@pschuermann97 What about this PR?

**Comment by pschuermann97** [comment]
this is an old attempt to fix the tests for the incentive service in presence of offline double-spending protection (problem was that in the unit tests for the incentive service, no dsp service instance was running, so all tests involving the spend-deduct protocol failed).

In this PR, I implemented a flag to turn the double-spending protection off where it was not needed.

However, you conviced me that this is a bad idea since it introduces an attack vector for the sake of testing.

Long story short, this PR and branch are no longer needed.
